### PR TITLE
Change how the cargo enroute age is calculated

### PR DESCRIPTION
**Pull request by YetiWiz**
This PR aims to deal with the cargo transfer (original) bug: https://github.com/OpenLoco/OpenLoco/issues/130

An example of how it works:
At a station there are 60 tons of coal (25 days old).
20 tons of coal (5 days old) are added.
The days for the 'enroute' age are added up: 25 + 5 = 30 days
Then, (if there was previous cargo) the result is reduced using the following formula:
`enrouteAge = enrouteAge * (1 - addedQuantity / summedQuantity)`

So the new age would be 30 * (1 - 20 / 80) = 20 days
The new 'enroute' age for the 80 tons is now 20 days.

This should make transferring cargo from, say, a ship to a train which then finishes the delivery, finally viable.

### CODE REVIEW

File: src/OpenLoco/Vehicles/VehicleHead.cpp [hunk 1]
```diff
@@ -2676,9 +2676,16 @@ namespace OpenLoco::Vehicles
 
                 break;
             }
+
+            bool stationHadPreviousCargo = cargoStats.quantity != 0;
             cargoStats.quantity = Math::Bound::add(cargoStats.quantity, cargo.qty);
             station->updateCargoDistribution();
-            cargoStats.enroute_age = std::max(cargoStats.enroute_age, cargo.numDays);
+            cargoStats.enrouteAge += cargo.numDays;
```
Comment (AaronVanGeffen) [inline src/OpenLoco/Vehicles/VehicleHead.cpp:2683]: We risk an overflow here, right?

Comment (YetiWiz) [reply src/OpenLoco/Vehicles/VehicleHead.cpp]: Oh true, I've changed it to use the bound like above, idk how I haven't noticed that before.

File: src/OpenLoco/Vehicles/VehicleHead.cpp [hunk 2]
```diff
@@ -2676,9 +2676,16 @@ namespace OpenLoco::Vehicles
...
             }
+
+            bool stationHadPreviousCargo = cargoStats.quantity != 0;
             cargoStats.quantity = Math::Bound::add(cargoStats.quantity, cargo.qty);
             station->updateCargoDistribution();
-            cargoStats.enroute_age = std::max(cargoStats.enroute_age, cargo.numDays);
+            cargoStats.enrouteAge = Math::Bound::add(cargoStats.enrouteAge, cargo.numDays);
+            if (stationHadPreviousCargo)
+            {
+                cargoStats.enrouteAge *= 1 - static_cast<float>(cargo.qty) / cargoStats.quantity;
```
Comment (duncanspumpkin) [inline src/OpenLoco/Vehicles/VehicleHead.cpp:2686]: never use floats in this code base. Instead multiply both sides by a constant and then divide out the constant.
e.g.

`((cargo.qty << 8) / cargoStats.quantity) >> 8` 
You'll need to have a think about if it needs higher accuracy and if it needs to be promoted to 64bit.

Comment (duncanspumpkin) [reply src/OpenLoco/Vehicles/VehicleHead.cpp:2686]: Urgh just noticed what you are doing will need to be more complex than that.
Do:
```cpp
const auto multiplier = (1 << 16) - (cargo.qty << 16) / cargoStats.quantity;
cargoStats.enrouteAge = (cargoStats.enrouteAge * multiplier) >> 16
```

Comment (YetiWiz) [reply src/OpenLoco/Vehicles/VehicleHead.cpp]: I see, thanks! I've left a comment to make it easier to understand what's happening.

**Comment by LeftofZen** [comment]
I agree with @YetiWiz here, that this is a bug and should be fixed. Trains shouldn't generate negative income just because the cargo passes through an intermediate station.

Duncan is correct (in his discord message) in that fixing this will affect game balance by making the game even easier because your trains will now generate more money, but a bug is a bug IMO and it should be fixed.

File: src/OpenLoco/Station.cpp [hunk 3]
```diff
@@ -578,7 +578,11 @@ namespace OpenLoco
             {
                 if (cargo.quantity != 0 && cargo.origin != id())
                 {
-                    cargo.enroute_age = std::min(cargo.enroute_age + 1, 255);
+                    cargo.enrouteAge = std::min(cargo.enrouteAge + 1, 255);
+                }
+                else
+                {
+                    cargo.enrouteAge = 0;
```
Comment (duncanspumpkin) [inline src/OpenLoco/Station.cpp:585]: Take me through why this change? To me this looks like it forces all enroute cargo to have zero age

Comment (duncanspumpkin) [reply src/OpenLoco/Station.cpp:585]: Never mind i understand now you are just tidying it up

File: src/OpenLoco/Vehicles/VehicleHead.cpp [hunk 4]
```diff
@@ -2676,9 +2676,18 @@ namespace OpenLoco::Vehicles
...
             }
+
+            const bool stationHadPreviousCargo = cargoStats.quantity != 0;
             cargoStats.quantity = Math::Bound::add(cargoStats.quantity, cargo.qty);
             station->updateCargoDistribution();
-            cargoStats.enroute_age = std::max(cargoStats.enroute_age, cargo.numDays);
+            cargoStats.enrouteAge = Math::Bound::add(cargoStats.enrouteAge, cargo.numDays);
+            if (stationHadPreviousCargo)
+            {
+                // enrouteAge = enrouteAge * (1 - addedQuantity / summedQuantity)
```
Comment (duncanspumpkin) [inline src/OpenLoco/Vehicles/VehicleHead.cpp:2686]: Add a comment that this is calculating the weighted average of the two cargo sources and that this is a change from vanilla.

**Comment by AaronVanGeffen** [comment]
This warrants a changelog entry :)

**Comment by duncanspumpkin** [comment]
I did a divergence test and since the lets play never uses intermediate stations it doesn't trigger. Oh well.
